src/charm.py

```diff
@@ -57,17 +58,36 @@ def __init__(self, *args: typing.Any):
         self.framework.observe(
             self.on[constants.PEER].relation_departed, self._on_peer_relation_departed
         )
-        self.peer_relation = self.model.get_relation(constants.PEER)
+        self.framework.observe(
+            self.on[constants.PEER].relation_joined, self._on_peer_relation_joined
+        )
 
     def _on_reload_bind(self, _: events.ReloadBindEvent) -> None:
         """Handle periodic reload bind event."""
+        # Reloading is used to take new configuration into account
         try:
             relation_data = self.dns_record.get_remote_relation_data()
         except ValueError as err:
             logger.info("Validation error of the relation data: %s", err)
             return
-        if self.bind.has_a_zone_changed(relation_data):
-            self.bind.update_zonefiles_and_reload(relation_data)
+        topology = self._topology()
+        if self.bind.has_a_zone_changed(relation_data, topology):
+            self.bind.update_zonefiles_and_reload(relation_data, topology)
+
+    def _on_peer_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
```

Comment:
Don't you plan to unit test this method?

Reply:
I don't think that it is useful. This function is purely charm behavior, what it calls should be covered by ops tests.
I check that the logic is correct through the integration tests